components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt

```diff
@@ -0,0 +1,311 @@
+        appWidgetManager.updateAppWidget(appWidgetId, views)
+    }
+
+    private fun createRemoteViews(
+        context: Context,
+        layout: Int,
+        textSearchIntent: PendingIntent,
+        voiceSearchIntent: PendingIntent?,
+        text: String?
+    ): RemoteViews {
+        return RemoteViews(context.packageName, layout).apply {
+            setSearchWidgetIcon(context)
+            setMicrophoneIcon(context)
+            when (layout) {
+                R.layout.mozac_search_widget_extra_small_v1,
+                R.layout.mozac_search_widget_extra_small_v2,
+                R.layout.mozac_search_widget_small_no_mic -> {
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_new_tab,
+                        textSearchIntent
+                    )
+                }
+                R.layout.mozac_search_widget_small -> {
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_new_tab,
+                        textSearchIntent
+                    )
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_voice,
+                        voiceSearchIntent
+                    )
+                }
+                R.layout.mozac_search_widget_medium,
+                R.layout.mozac_search_widget_large -> {
+
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_new_tab,
+                        textSearchIntent
+                    )
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_voice,
+                        voiceSearchIntent
+                    )
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_new_tab_icon,
+                        textSearchIntent
+                    )
+                    setTextViewText(R.id.mozac_button_search_widget_new_tab, text)
+
+                    // Unlike "small" widget, "medium" and "large" sizes do not have separate layouts
+                    // that exclude the microphone icon, which is why we must hide it accordingly here.
+                    if (voiceSearchIntent == null) {
+                        setViewVisibility(R.id.mozac_button_search_widget_voice, View.GONE)
+                    }
+                }
+            }
+        }
+    }
+
+    private fun RemoteViews.setMicrophoneIcon(context: Context) {
+        setImageView(
+            context,
+            R.id.mozac_button_search_widget_voice,
+            config.searchWidgetMicrophoneResource
+        )
+    }
+
+    private fun RemoteViews.setSearchWidgetIcon(context: Context) {
+        setImageView(
+            context,
+            R.id.mozac_button_search_widget_new_tab_icon,
+            config.searchWidgetIconResource
+        )
+        val appName = context.getString(R.string.app_name)
```

Comment:
Should this be in the config class too? R.string.app_name isn't a protected ID reference so apps might have even used `R.string.application_name` in their app and our overriding might not have worked.

Reply:
Done